Python/chapter01/1.3 - URLify/miguel_1.3_sol.py

```diff
@@ -24,22 +24,21 @@ def urlify(s: str, true_length: int) -> str:
     # Below link goes over string concat efficiency in python (I will use method 4
     # https://waymoot.org/home/python_string/
     # each space will make us replace it (with '%', then add 2 characters ('2', '0')
-    result = ['\x00'] * len(s)*2  # extra buffer space
-    i = 0
+    # must count spaces first to help us find out the size of the
+    result = ['\x00'] * true_length * 3
     space_count = 0
-    added_chars_per_space = 2
-    for c in s:
-        if i == true_length + added_chars_per_space * space_count:
-            break
-        if c == ' ':
-            result[i] = '%'
-            result[i+1] = '2'
-            result[i+2] = '0'
-            i += 3
+    chars_per_space = 2
+    idx = 0
+    for x in range(true_length):
+        if s[idx - space_count * chars_per_space] == ' ':
```

Comment:
remember that `s` is inaccessible in the challenge mode.

Reply:
I'm not sure I understand.  How else would I get the character values of the argument s?

Reply:
actually let me formalize the challenge to avoid confusion:

```python3
from typing import List

def urlify_challenge(s: str) -> str:
  """In-place version of urlify."""
  buf = ['\x00'] * (len(s) * 3)
  for i, c in enumerate(s):
    buf[i] = c
  def _challenge(buf: List[str], original_length: int) -> None:
    # Your code here. Challenge rules:
    # * You can only reference |buf| and |original_length|, not |s|.
    # * You can only allocate O(1) additional memory
    # * You need to modify |buf| in-place so that the right answer is returned.
  _challenge(buf, len(s))
  return ''.join(buf).rstrip('\x00')
```

you can get the character values through the buffer.

Reply:
Oh ok, I see what you mean now.  Will give that a go:)

Reply:
Lol I mistakenly still used s in my initial _challenge attempt, idk how that missed that.  My negligence.

Has been updated to modify buf in place and NOT reference s.